Bugfix: backdrop not shared in different elements, causing inconsistent closing

zinkkrysty opened the pull request:
Bug described in #73. This pull request adds the failing tests. 

My approach in #74 is flawed and therefore re-created the tests without the solution. Let's discuss what a reasonable approach to the solution would be. I think that @valdrinkoshi is right, we shouldn't have multiple backdrops, we should only allow one backdrop and that should be managed by the `iron-overlay-manager`.

**EDIT** I've now commited the solution, much more elegant and straightforward: don't allow more than 1 instance of `iron-overlay-backdrop` and let the manager handle the creation of that.

valdrinkoshi commented on iron-overlay-behavior.html:
```diff
@@ -155,11 +155,8 @@
-    },
-
     ready: function() {
+      this._backdrop = this._manager.getBackdropInstance();
```
you could completely remove `this._backdrop` from here, and just return `this.manager.getBackdropInstance()` in `get backdropElement()`

valdrinkoshi commented on iron-overlay-manager.html:
```diff
@@ -106,6 +106,13 @@
       }
     },
 
+    getBackdropInstance: function() {
```
you could make this a getter, in the same fashion as in `iron-overlay-behavior`
```javascript
// ...
_backdropInstance: document.createElement('iron-overlay-backdrop'),

get backdropElement() {
   return this._backdropInstance;
},
// ...

zinkkrysty replied:
great idea!

zinkkrysty replied:
I'm just wondering, if it's slightly more efficient to only create the backdrop element in case we need it? lazy-load vs eager-load ...

valdrinkoshi replied:
you're right, you can do the lazy loading within the `get backdropElement()`

valdrinkoshi commented on test/iron-overlay-behavior.html:
```diff
@@ -421,6 +434,45 @@
+          assert.equal(overlays[0].backdropElement, overlays[1].backdropElement);
+        });
+
+        test('multiple overlays close the corrent backdrop', function(done) {
```
small typo, "corrent" should be "current"

zinkkrysty replied:
I actually meant "correct". But since we are only talking about one single backdrop now maybe it should just be "the backdrop"

valdrinkoshi replied:
fair enough :+1:

valdrinkoshi commented on test/iron-overlay-behavior.html:
```diff
@@ -421,6 +434,45 @@
+        });
+
+        test('multiple overlays close the corrent backdrop', function(done) {
+          runAfterOpen(overlays[0], function () {
```
Consider adding a method `runAfterClose` which is the same as `runAfterOpen`, but just adds a listener for `iron-overlay-close` instead of `iron-overlay-opened`.
Also, notice that `runAfterOpen` will execute the callback after 1ms.
With that this code could become something like this:
```javascript
runAfterOpen(overlays[0], function() {
  overlays[0].close();
  runAfterOpen(overlays[1], function() {
    runAfterClose(overlays[1], function() {
       // your tests here!
    });
  });
});

valdrinkoshi commented on iron-overlay-behavior.html:
```diff
@@ -155,11 +155,8 @@
     },
 
-    registered: function() {
-      this._backdrop = document.createElement('iron-overlay-backdrop');
```
we should also move the import to `iron-overlay-manager` (line 14)
```html
<link rel="import" href="iron-overlay-backdrop.html">

zinkkrysty replied:
makes sense I'll do that and amend the previous commit with fixes

valdrinkoshi replied:
oh, and we can also remove it from here (the line 14), since the import of `iron-overlay-manager` will import the backdrop :)

zinkkrysty replied:
I thought I did that, but somehow I didn't commit :) I'll do it now.

valdrinkoshi replied:
this was the suggestion that causes the tests to fail :x
Never mind it, the import should be done here, in `iron-overlay-behavior` since in here we use the methods of `iron-overlay-backdrop`, and should not be done in `iron-overlay-manager`

valdrinkoshi commented on iron-overlay-manager.html:
```diff
@@ -22,6 +23,8 @@
 
     _backdrops: [],
 
+    _backdropElement: document.createElement('iron-overlay-backdrop'),
```
@zinkkrysty tests are failing because of this line. If you do lazy loading on the getter solves it

valdrinkoshi replied:
the reason why this is failing is because the manager is a singleton, an object, and here we add a key. To access to it we should do `Polymer.IronOverlayManager._backdropElement` in the getter (ugly).
Much better to lazy load by checking if `this._backdropInstance` exists.
```javascript
   get backdropElement() {
      if (!this._backdropElement) {
        this._backdropElement = document.createElement('iron-overlay-backdrop');
      }
      return this._backdropElement;
    },

zinkkrysty replied:
thanks I was trying to run the tests locally with wct - that thing doesn't want to work with my setup, but you've beat me to it figuring out the issue. I'll make this fix and commit once more

zinkkrysty replied:
Hmm there is a problem now, a circular dependency issue I think: the manager imports the backdrop and the backdrop imports the manager, I guess that's why I get `Uncaught TypeError: Cannot read property 'getBackdrops' of undefined`

zinkkrysty replied:
Yeah I can confirm it. If I move the `backdrop` import back into `iron-overlay-behavior` the tests pass

valdrinkoshi replied:
Oh bummer...
Well, thinking of it again, it makes sense for `iron-overlay-behavior` to do the import, after all he's the one that uses the backdropElement's methods. Sorry for the confusion, my previous suggestion introduced this issue :disappointed:

zinkkrysty replied:
Well that's fine, it made sense for me as well. But I think this is a bit of code smell. I mean there shouldn't be circular dependency between 2 elements, should there?

zinkkrysty commented:
Thanks for the suggestions and the help @valdrinkoshi , it's nice to see this bug finally squashed :) :tada:

valdrinkoshi commented on iron-overlay-behavior.html:
```diff
@@ -11,8 +11,8 @@
 <link rel="import" href="../iron-resizable-behavior/iron-resizable-behavior.html">
-<link rel="import" href="iron-overlay-backdrop.html">
 <link rel="import" href="iron-overlay-manager.html">
+<link rel="import" href="iron-overlay-backdrop.html">
```
would you mind reverting this as before? :angel: 
(since there is no particular reason to be moved as last import)

zinkkrysty replied:
hmm, sure. My reasoning was that the backdrop needs the manager before but if I think about it, the backdrop actually imports the manager, so should be fine. I'll do that now.

valdrinkoshi commented:
Awesome job @zinkkrysty! I've let a minor comment, & I have a last request: could you squash your commits into one (so we have a clean history in our repo) ?

Overall LGTM, I defer the final :+1: and merge to @cdata :)

zinkkrysty commented:
Ok thanks! I'll squash the commits. I wasn't sure if you needed the history

valdrinkoshi commented:
Thanks @zinkkrysty, LGTM :+1: 
@cdata, ball in your court :soccer:

valdrinkoshi commented:
@zinkkrysty would you mind rebase this one? For that getter on `iron-overlay-manager`, you can use [`Object.defineProperty`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/defineProperty)

zinkkrysty commented:
Sure, I'll give it a go now

zinkkrysty commented:
Bummer, I've implemented the getter like this:

```
  Object.defineProperty(Polymer.IronOverlayManagerClass.prototype, "backdropElement", {
    get: function() {
      if (!this._backdropElement) {
        this._backdropElement = document.createElement('iron-overlay-backdrop');
      }
      return this._backdropElement;
    }
  });
```

But get a failing test at: 

```
assert.lengthOf(document.querySelectorAll('iron-overlay-backdrop'), 0); // It's 1 instead
```

Meaning that somehow the backdrop is not removed from the document when the overlays are closed.

I can push to this branch, but I already rebased and I wanted to show that the tests were indeed passing and my fix was working previously.

zinkkrysty commented:
Could the removal of the `completeBackdrop` method be the source of this regression? I guess that was the one that was removing the backdrop once it wasn't used anymore.

https://github.com/PolymerElements/iron-overlay-behavior/commit/61928b26e72d1771c6d96021cb04c062599d2504#diff-5e1626c54b6ff0c467b99e2307ad8af4L352

zinkkrysty commented:
Ah, of course, the new code closes the backdrop after transition ends, that's why if I add a timeout of 1s to the test, it passes. 

Any ideas on how to check the presence of the backdrop in the DOM after the `transitionEnd` callback, or do you think it's better to remove that assertion, as it's being tested for elsewhere?

valdrinkoshi commented:
You can manually call:
```javascript
overlay[1].close();
overlay[1].backdropElement.complete();
```
So that you don't need to wait for animations to be over

zinkkrysty commented:
Ok, I guess that defeats the purpose. I wanted to make sure that `complete()` is called after the backdrops are closed, but that is probably subject for another test. I'll remove that line and push the changes.

cdata commented:
This LGTM :+1: Thanks for the contribution!
